src/Google.Events.Protobuf.Tests/ExtensionAttributeTest.cs

```diff
@@ -0,0 +1,30 @@
+﻿// Copyright 2023, Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Events.Protobuf.Cloud.Functions.V2;
+using Xunit;
+
+namespace Google.Events.Protobuf.Tests;
+
+public class ExtensionAttributeTest
+{
+    [Fact]
+    public void ExtensionAttributeFromDeserializedEvent()
+    {
+        var bytes = TestResourceHelper.LoadBytes("structured-mode-body-with-extension.json");
+        var formatter = new ProtobufJsonCloudEventFormatter<FunctionEventData>();
+        var evt = formatter.DecodeStructuredModeMessage(bytes, null, null);
+        Assert.Equal("sample-function", evt[ExtensionAttributes.Function]);
```

Comment:
If Function is an extension that applies to another event, what's the error message?
Maybe a test for that, if only for documenting the error and trying to improve if needed, could be fine.

I know there's already a TODO on the generation code to take into account the event type the attributes apply to, which would help a lot here. I've left a comment there about that TODO maybe being something to do now?

Also, maybe consider making these actual extension methods: `evt.GetFunction();`? The Extension name predisposed me to finding extension methods.

Reply:
> If Function is an extension that applies to another event, what's the error message?

There won't be an error - it'll just return null (IIRC, anyway). I'll add a test for that, thanks.

> I know there's already a TODO on the generation code to take into account the event type the attributes apply to, which would help a lot here. I've left a comment there about that TODO maybe being something to do now?

Hmm. I don't think it's worth changing how the attributes are generated (in terms of API surface)... partly because there's a problem here in terms of the same event message (e.g. FunctionEventData) being used for multiple event types.

However, I could:

- Only generate them when we know of at least one event type
- Generate documentation saying which event types they're used for

> Also, maybe consider making these actual extension methods: evt.GetFunction();? The Extension name predisposed me to finding extension methods.

Right, that's CloudEvent terminology clashing with C# terminology. I don't think I want to add extension methods at the moment, partly because it's so easy to get into collisions if you use multiple namespaces. But we could add that later, if we get customer feedback.

Reply:
Yep, all this sounds good, including documentation for the extension attributes. And thanks for the test.